dao-voting/tests/harness.rs

```diff
@@ -0,0 +1,1156 @@
+
+            vote(&user.dao_voting, true, 1, asset_amount / 4).await;
+
+            assert_eq!(
+                user_balance(&user.dao_voting, Identity::Address(user.wallet.address())).await,
+                4
+            );
+
+            assert_eq!(
+                user_votes(
+                    &user.dao_voting,
+                    Identity::Address(user.wallet.address()),
+                    1
+                )
+                .await,
+                Votes {
+                    yes_votes: 2,
+                    no_votes: 0,
+                }
+            );
+        }
+    }
+
+    mod revert {
+        use super::*;
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_on_invalid_proposal_id() {
+            let (_gov_token, _gov_token_id, _deployer, user, _asset_amount) = setup().await;
+            vote(&user.dao_voting, true, 0, 10).await;
+        }
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_on_zero_vote_amount() {
+            let (_gov_token, gov_token_id, deployer, user, _asset_amount) = setup().await;
+            constructor(&deployer.dao_voting, gov_token_id).await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user.dao_voting, 10, 10, proposal_transaction.clone()).await;
+            vote(&user.dao_voting, true, 0, 0).await;
+        }
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_on_expired_proposal() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+            constructor(&deployer.dao_voting, gov_token_id).await;
+
+            mint(
+                &deployer.gov_token.as_ref().unwrap(),
+                asset_amount,
+                user.wallet.address(),
+            )
+            .await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user.dao_voting, 1, 1, proposal_transaction.clone()).await;
+
+            let call_params = CallParameters::new(
+                Some(asset_amount),
+                Some(AssetId::from(*gov_token_id)),
+                Some(100_000),
+            );
+            deposit(&user.dao_voting, call_params).await;
+            vote(&user.dao_voting, true, 0, asset_amount / 4).await;
+        }
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_on_vote_amount_greater_than_balance() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+            constructor(&deployer.dao_voting, gov_token_id).await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user.dao_voting, 10, 10, proposal_transaction.clone()).await;
+            vote(&user.dao_voting, true, 10, asset_amount).await;
+        }
+
+        // TODO add test for reverting on a failed proposal call
```

Comment:
I'd move this comment down a bit so that it is near the revert of the execute suite rather than the voting suite